File: universal/src/main/java/com/github/fernthedev/INewEntityRegistry.java
package com.github.fernthedev;

import com.github.fernthedev.fernutils.thread.ThreadUtils;
import com.github.fernthedev.fernutils.thread.multiple.TaskInfoList;
import com.github.fernthedev.universal.EntityID;
import com.github.fernthedev.universal.GameObject;
import com.github.fernthedev.universal.UniversalHandler;
import com.github.fernthedev.universal.entity.EntityPlayer;
import com.google.common.base.Stopwatch;
import lombok.Getter;
import lombok.NonNull;

import java.util.*;
import java.util.concurrent.ExecutorService;
import java.util.stream.Collectors;

public abstract class INewEntityRegistry extends TickRunnable {

    @Getter
    protected Map<@NonNull UUID, @NonNull GameObject> gameObjects = Collections.synchronizedMap(new HashMap<>());

    public void addEntityObject(@NonNull GameObject gameObject) {
        // For interpolation
        float prevX = gameObject.getPrevX();
        float prevY = gameObject.getPrevY();

        if (gameObjects.containsKey(gameObject.getUniqueId()) && gameObjects.get(gameObject.getUniqueId()).hashCode() != gameObject.hashCode()) {
            prevX = gameObject.getX();
            prevY = gameObject.getY();
        }

        gameObject.setPrevX(prevX);
        gameObject.setPrevY(prevY);

        gameObjects.put(gameObject.getUniqueId(), gameObject);
    }


    public void removeEntityObject(UUID gameObject) {
        gameObjects.remove(gameObject);
    }

    public void removeEntityObject(GameObject gameObject) {
        removeEntityObject(gameObject.getUniqueId());
    }

    public void collisionCheck(EntityPlayer universalPlayer) {
        List<GameObject> gameObjectsCheck = copyGameObjectsAsList();

        for (GameObject tempObject : gameObjectsCheck) {
            if (universalPlayer.getBounds().intersects(tempObject.getBounds())) {
                if (tempObject.getEntityId() == EntityID.ENEMY) {
                    //COLLISION CODE
                    universalPlayer.setHealth(universalPlayer.getHealth() - 2);
                }


                if (tempObject.getEntityId() == EntityID.COIN) {
                    universalPlayer.setCoin(universalPlayer.getCoin() + 1);
                    removeEntityObject(tempObject);
                    UniversalHandler.getIGame().getLoggerImpl().info("COllision checking! COIN");
                    // this.handler.removeObject(tempObject);
                }
            }
        }
    }

    public void tick() {
        List<GameObject> objects = copyGameObjectsAsList();


        TaskInfoList taskInfoList = ThreadUtils.runForLoopAsync(objects, tempObject -> {
            tempObject.tick();

            clampAndTP(tempObject);

            if (tempObject instanceof EntityPlayer)
                collisionCheck((EntityPlayer) tempObject);
        });

        Stopwatch stopwatch = Stopwatch.createStarted();
        taskInfoList.runThreads(getExecutorService());

        // Wait for all objects to finish ticking
        taskInfoList.awaitFinish(1);
        stopwatch.stop();
    }



    protected String clampAndTP(GameObject gameObject) {

        String result = "";

        float maxX = (float) UniversalHandler.WIDTH - (float) gameObject.getWidth() - gameObject.getWidth()/2f;
        float maxY = (float) UniversalHandler.HEIGHT - (float) gameObject.getHeight()*2f - gameObject.getWidth()/4f;

        float newX = GameMathUtil.clamp(gameObject.getX(), 0, maxX);
        float newY = GameMathUtil.clamp(gameObject.getY(), 0, maxY);

        if (newX != gameObject.getX() && gameObject instanceof EntityPlayer) {
            gameObject.setVelX(0);
            gameObject.setX(newX);
            result += "X was clamped " + newX;
        }

        if (newY != gameObject.getY() && gameObject instanceof EntityPlayer) {
            result += " Y was clamped " + newY;
            gameObject.setY(newY);
            gameObject.setVelY(0);
        }

        if (result.isEmpty())
            result = null;

        return result;
    }


    public Map<UUID, GameObject> copyGameObjectsAsMap() {
        return new HashMap<>(getGameObjects());
    }

    public List<GameObject> copyGameObjectsAsList() {
        return new ArrayList<>(getGameObjects().values())
                .parallelStream()
                .collect(Collectors.toList());
    }

    public Map<UUID, Integer> getObjectsAndHashCode() {
        return getObjectsAndHashCode(gameObjects.keySet());
    }

    public Map<UUID, Integer> getObjectsAndHashCode(Set<UUID> uuids) {
        Map<UUID, Integer> map = new HashMap<>();

        copyGameObjectsAsMap().forEach((uuid, gameObjectIntegerPair) -> {
            if(uuids.contains(uuid))
                map.put(uuid, gameObjectIntegerPair.hashCode());
        });

        return map;
    }




    protected abstract ExecutorService getExecutorService();
}

File: universal/src/main/java/com/github/fernthedev/INewEntityRegistry.kt
package com.github.fernthedev

import com.github.fernthedev.universal.EntityID
import com.github.fernthedev.universal.GameObject
import com.github.fernthedev.universal.Location
import com.github.fernthedev.universal.UniversalHandler
import com.github.fernthedev.universal.entity.EntityPlayer
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import kotlin.system.measureTimeMillis

abstract class INewEntityRegistry : TickRunnable {
//    private val tickThreadGroup: Array<Thread>
//    init {
//        val size = max(Runtime.getRuntime().availableProcessors() / 2, 1)
//        tickThreadGroup = Array(size, i -> {
//            Thread()
//        })
//
//    }

    private val _gameObjects: MutableMap<UUID, GameObject> = ConcurrentHashMap()
    private val _previousLocations: MutableMap<UUID, Location> = ConcurrentHashMap()

    open val gameObjects: Map<UUID, GameObject>
        get() = _gameObjects
    val previousLocations: Map<UUID, Location>
        get() = _previousLocations

    fun clearEntities() {
        _gameObjects.clear()
        _previousLocations.clear()
    }

    fun addEntityObject(gameObject: GameObject) {
//        if (gameObjects.containsKey(gameObject.uniqueId)) {
//            throw IllegalArgumentException("GameObject ${gameObject.uniqueId} already exists")
//        }

        // For interpolation
        _gameObjects[gameObject.uniqueId] = gameObject
        _previousLocations[gameObject.uniqueId] = gameObject.location.copy()
    }

    fun removeEntityObject(gameObject: UUID) {
        _gameObjects.remove(gameObject)
        _previousLocations.remove(gameObject)
    }

    fun removeEntityObject(gameObject: GameObject) {
        removeEntityObject(gameObject.uniqueId)
    }

    open fun collisionCheck(universalPlayer: EntityPlayer) {
        for ((_, tempObject) in gameObjects) {
            if (!universalPlayer.bounds.intersects(tempObject.bounds)) continue

            if (tempObject.entityId == EntityID.ENEMY) {
                //COLLISION CODE
                universalPlayer.health = universalPlayer.health - 2
            }

            if (tempObject.entityId == EntityID.COIN) {
                universalPlayer.coin = universalPlayer.coin + 1
                removeEntityObject(tempObject)
            }
        }
    }

    override suspend fun tick() = coroutineScope {
        val ms = measureTimeMillis {

            val jobs = gameObjects.map { (_, obj) ->
                val job = launch {
                    obj.tick()
                    clampAndTP(obj)
                    if (obj is EntityPlayer) collisionCheck(obj)
                }

                job
            }.toList()

            jobs.forEach { job ->
                job.join()
            }

        }

        if (ms > 10) {
            println("Tick took $ms")
        }
    }

    protected open fun clampAndTP(gameObject: GameObject) {
        val location = gameObject.location
        location.x += gameObject.velX
        location.y += gameObject.velY

        val maxX: Float = UniversalHandler.WIDTH.toFloat() - gameObject.width - gameObject.width / 2f
        val maxY: Float = UniversalHandler.HEIGHT.toFloat() - gameObject.height * 2f - gameObject.width / 4f

        val newX: Float = GameMathUtil.clamp(gameObject.location.x, 0F, maxX)
        val newY: Float = GameMathUtil.clamp(gameObject.location.y, 0F, maxY)

        if (gameObject is EntityPlayer) {
            if (newX != gameObject.location.x) {
                gameObject.velX = 0F
            }
            if (newY != gameObject.location.y) {
                gameObject.velY = 0F
            }
        } else {
            if (location.x <= 0 || location.x >= maxX) gameObject.velX *= -1f
            if (location.y <= 0 || location.y >= maxY) gameObject.velY *= -1f
        }

        val prevLoc = previousLocations[gameObject.uniqueId]
        if (prevLoc != null) {
            prevLoc.x = gameObject.location.x
            prevLoc.y = gameObject.location.y
        }

        gameObject.location.x = newX
        gameObject.location.y = newY
    }
}